scripts/release-asm/merge

```diff
@@ -16,3 +17,48 @@ $(cat "$@")
 
 main "\${@}"
 EOF
+
+  # linter to make sure all variables are either read-only globals or locals
+  lint "${out}"
+}
+
+lint() {
+  local FILE; FILE="${1}"
+  local FAILED; FAILED=0
+
+  while read -r VARIABLE; do
+    VARIABLE=${VARIABLE%=}
+    if ! grep -Eq "local([[:space:]A-Za-z0-9_\-]+)?${VARIABLE}(;|$|=|[[:space:]])" "${FILE}" \
+    && ! grep -Eq "readonly ${VARIABLE}(;|$)" "${FILE}"; then
+      echo "FAILED: ${VARIABLE} is not marked as local or readonly!"
+      FAILED=1
+    fi
+  done <<EOF
+$(get_all_variables "${FILE}")
+EOF
+
+  if [[ "${FAILED}" -eq 1 ]]; then
+    false
+  fi
+}
+
+get_all_variables() {
+  local FILE; FILE="${1}"
+  local VARIABLE_PATTERN; VARIABLE_PATTERN="[A-Za-z0-9_\-]\+="
+  cat "${FILE}" | grep "${VARIABLE_PATTERN}" | grep -vE "$(skip_pattern)" | grep -o "${VARIABLE_PATTERN}" | uniq
+}
+
+skip_pattern() {
+  local PATTERN
+
+  PATTERN="[[:space:]]\-"
+  PATTERN="${PATTERN}|\-\-"
+  PATTERN="${PATTERN}|shellcheck"
+  PATTERN="${PATTERN}|^#"
+  PATTERN="${PATTERN}|echo"
+  PATTERN="${PATTERN}|.split\("
```

Comment:
Can you add comments for some of these justifying why we're skipping them? The `^#` I understand (although I feel like it should probably be `^[[:space:]]*#`), but some of them seem like perfectly valid cases. e.g. we shouldn't skip

`VAR=VAL || echo "Failed to set"`

Reply:
Chatted offline to lint ALL_CAPs only.